File: c++/9466.cpp
#include <iostream>
#include <string.h>

using namespace std;

int where[100001];
int cycle[100001];
bool visit[100001];
int T,N,cnt;

void dfs(int start){
    visit[start]=true;
    int next = where[start];
    if(visit[next]==false){
        dfs(next);
    }
    else if(cycle[next]==0){
        for(int i=next ; i!=start ; i=where[i]){
            cnt++;
        }
        cnt++;
    }
    cycle[start]=1;
}

int main(){
    scanf("%d",&T);
    while(T-->0){
        scanf("%d",&N);
        for(int i=1 ; i<=N ; i++){
            scanf("%d",&where[i]);
        }
        cnt=0;
        for(int i=1 ; i<=N ; i++){
            if(visit[i]==false){
                dfs(i);
            }
        }
        printf("%d\n",N-cnt);

        memset(where,0,sizeof(where));
        memset(cycle,0,sizeof(cycle));
        memset(visit,false,sizeof(visit));
    }

    return 0;
}

File: python/9466.py
import sys

sys.setrecursionlimit(10**7)

T = int(sys.stdin.readline())


def dfs(num_list, cycle, result, start):
    result[start] = 1
    after = num_list[start]
    if result[after] == 1:
        if after in cycle:
            return cycle[cycle.index(after):]
        return
    else:
        cycle.append(after)
        return dfs(num_list, cycle, result, after)


def solution():
    N = int(sys.stdin.readline())
    people = [0] + list(map(int, sys.stdin.readline().split()))
    result = [0]*(N+1)
    cnt = 0
    for i in range(1, N+1):
        if result[i] == 0:
            temp = [i]
            cycle = dfs(people, temp, result, i)
            if cycle:
                cnt += len(cycle)

    print(N-cnt)


while T > 0:
    solution()
    T -= 1
